### Man Wayne Hui

#### Week 5 Assignment 4

Your task in this week’s assignment is to answer three questions:
1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather?

In [136]:
import pandas as pd

# Load datasets directly from the GitHub raw URLs
airports_url = "https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/airports.csv"
weather_url = "https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/weather.csv"

airports = pd.read_csv(airports_url)
weather = pd.read_csv(weather_url)


### Importing data
 * Import pandas: Loads the data manipulation library.
 * Define URLs: Points directly to the raw CSV files for airports and weather on GitHub.
 * Read CSVs: Uses `pd.read_csv()` to download and convert the data into pandas DataFrames for analysis.

In [137]:
airports.head()

,faa,name,lat,lon,alt,tz,dst,tzone
0,04G,Lansdowne Airport,41.130472,-80.619583,1044,-5,A,America/New_York
1,06A,Moton Field Municipal Airport,32.460572,-85.680028,264,-6,A,America/Chicago
2,06C,Schaumburg Regional,41.989341,-88.101243,801,-6,A,America/Chicago
3,06N,Randall Airport,41.431912,-74.391561,523,-5,A,America/New_York
4,09J,Jekyll Island Airport,31.074472,-81.427778,11,-5,A,America/New_York


Inspect the first 5 rows of the `airports` dataset using `.head()` to verify that the data.

In [138]:
weather.head()

,origin,year,month,day,hour,temp,dewp,humid,wind_dir,wind_speed,wind_gust,precip,pressure,visib,time_hour
0,EWR,2013,1,1,1,39.02,26.06,59.37,270.0,10.35702,NaN,0.0,1012.0,10.0,2013-01-01T06:00:00Z
1,EWR,2013,1,1,2,39.02,26.96,61.63,250.0,8.05546,NaN,0.0,1012.3,10.0,2013-01-01T07:00:00Z
2,EWR,2013,1,1,3,39.02,28.04,64.43,240.0,11.50780,NaN,0.0,1012.5,10.0,2013-01-01T08:00:00Z
3,EWR,2013,1,1,4,39.92,28.04,62.21,250.0,12.65858,NaN,0.0,1012.2,10.0,2013-01-01T09:00:00Z
4,EWR,2013,1,1,5,39.02,28.04,64.43,260.0,12.65858,NaN,0.0,1011.9,10.0,2013-01-01T10:00:00Z


Inspect the first 5 rows of the `weather` dataset using `.head()` to verify that the data.

In [139]:
top_northern = airports.sort_values(by="lat", ascending=False).head(5)
top_northern[["faa", "name", "lat", "lon"]]

,faa,name,lat,lon
417,EEN,Dillant Hopkins Airport,72.270833,42.898333
230,BRW,Wiley Post Will Rogers Mem,71.285446,-156.766003
110,AIN,Wainwright Airport,70.638056,-159.994722
708,K03,Wainwright As,70.613378,-159.860350
152,ATK,Atqasuk Edward Burnell Sr Memorial Airport,70.467300,-157.436000


### Question 1 - What is the northernmost airport in the United States?
* Sort Values: Sort by `lat` with `ascending=False` to bring the highest latitudes to the top.
* Select Columns: `faa`, `name`, and `lat` for the top 5 results.

First I was going to assume that Dillant Hopkins Airport would be the northenrn most airport until I did a quick google search to find that the data is incorrect for EEN at 72.27° N.

**Answer:**
 Therefore the correct answer is **BRW (Wiley Post-Will Rogers Memorial Airport)** at 71.2854° N

In [140]:
airports.sort_values(by="lon", ascending=False).head(5)[["faa", "name", "lat", "lon"]]

,faa,name,lat,lon
1290,SYA,Eareckson As,52.712275,174.113620
942,MYF,Montgomery Field,32.475900,117.759000
396,DVT,Deer Valley Municipal Airport,33.411700,112.457000
417,EEN,Dillant Hopkins Airport,72.270833,42.898333
444,EPM,Eastport Municipal Airport,44.910111,-67.012694


## Question 2 - What is the easternmost airport in the United States?

* Sort Values: Sort by `lon` with `ascending=False` to bring the highest longitudes to the top.
* Select Columns: `faa`, `name`, `lat`, and `lon` for the top 5 results.

First, I looked up Eareckson As airport from the cordinates in the dataframe and it showed me the most west part of the United States. Then I saw airports like MYF (San Diego), DVT (Phoenix), and EEN at the top, but a quick search revealed they are data errors missing negative signs. After couple quick search of the 5 airports there are two technical answers.

**Answer:**

* `By Geographic Coordinates: Eareckson As (SYA)` on Shemya Island, is technically the easternmost U.S. airport because it crosses the 180° mark, its longitude flips to **+174.11° E**, placing it inside the Eastern Hemisphere.
* `By Conventional Geography (U.S. Map): Eastport Municipal Airport (EPM)` in Maine is the easternmost airport at **-67.01° W**, representing the furthest eastern point of the United States from looking at the US center point.

In [141]:
feb12_weather = weather[
    (weather["year"] == 2013) & 
    (weather["month"] == 2) & 
    (weather["day"] == 12)
]


feb12_weather.sort_values(by="wind_speed", ascending=False).head(5)[
    ["origin", "time_hour", "wind_speed"]
]

,origin,time_hour,wind_speed
1009,EWR,2013-02-12T08:00:00Z,1048.36058
18417,LGA,2013-02-12T07:00:00Z,23.01560
1018,EWR,2013-02-12T17:00:00Z,21.86482
18428,LGA,2013-02-12T18:00:00Z,21.86482
9717,JFK,2013-02-12T13:00:00Z,20.71404


In [142]:
clean_feb12 = feb12_weather[feb12_weather["wind_speed"] < 100]

wind_summary = clean_feb12.groupby("origin").agg(
    avg_wind_speed=("wind_speed", "mean"),
    max_wind_speed=("wind_speed", "max")
).reset_index()

wind_summary

,origin,avg_wind_speed,max_wind_speed
0,EWR,13.258987,21.86482
1,JFK,14.384750,20.71404
2,LGA,14.960140,23.01560


## Question 3 - On February 12th, 2013, which New York area airport had the windiest weather?

* Filter the `weather` dataset for February 12, 2013, and sort by `wind_speed` descending to view the top recorded wind speeds.

* At first glance, **EWR** records an impossible wind speed of **1,048.36 mph**, which is an obvious weather sensor glitch.

* Filter out readings over 100 mph (`wind_speed < 100`) to remove the corrupted data, then group by `origin` to calculate the daily average and peak wind speeds across the three airports.

* `.agg()` function compute multiple summary statistics (such as the average and maximum) across grouped columns all at once in a single clean table.

* `.groupby("origin")` splits the 24 hourly weather observations into separate groups for each airport (EWR, JFK, and LGA) so we can calculate summary statistics for it individually.

**Answer:**
After dropping the sensor error at EWR, **LaGuardia Airport (LGA)** had the windiest weather on February 12, 2013, with the highest average wind speed (~14.96 mph) and the highest sustained peak wind speed (~23.02 mph).